File: python/src/wdl_model/model/errors/wdl_exception.py
"""Aggregate exception carrying one or more WDL diagnostics."""

from __future__ import annotations

from typing import Iterable, List

from .wdl_error import WdlError


class WdlException(Exception):
    """Exception raised when parsing or validation collected diagnostics."""

    def __init__(self, errors: Iterable[WdlError]):
        self.errors: List[WdlError] = list(errors)
        super().__init__(self.toDebugMessage())

    def getErrors(self) -> List[WdlError]:
        """Return the collected diagnostics carried by the exception."""
        return self.errors

    def toDebugMessage(self) -> str:
        """Return a compact debug-oriented rendering of all collected diagnostics."""
        return "\n".join(error.toDebugMessage() for error in self.errors)

File: typescript/src/model/errors/wdl-exception.ts
/** Aggregate exception carrying one or more collected WDL diagnostics. */
import { WdlError } from './wdl-error.js';

export class WdlException extends Error {
  public readonly errors: readonly WdlError[];

  /** Creates an exception from a readonly diagnostic list. */
  public constructor(errors: readonly WdlError[]) {
    super(errors.map((error) => error.toDebugMessage()).join('\n'));
    this.errors = [...errors];
  }

  /** Returns the collected diagnostics carried by this exception. */
  public getErrors(): readonly WdlError[] {
    return this.errors;
  }

  /** Returns a compact debug-oriented rendering of all collected diagnostics. */
  public toDebugMessage(): string {
    return `Errors(${this.errors.map((error) => error.toDebugMessage()).join(',')})`;
  }
}
